src/Castle.Core/DynamicProxy/ProxyGenerator.cs

```diff
@@ -595,7 +595,10 @@ public object CreateInterfaceProxyWithTarget(Type interfaceToProxy, object targe
 						var result = Marshal.QueryInterface(iUnknown, ref interfaceId, out interfacePointer); // Increment the reference count
 						var isInterfacePointerNull = interfacePointer == IntPtr.Zero;
 						Marshal.Release(iUnknown); // Decrement the reference count
-						Marshal.Release(interfacePointer); // Decrement the reference count
+						if (!isInterfacePointerNull)
+						{
+							Marshal.Release(interfacePointer); // Decrement the reference count
+						}
 
 						if (result == 0 && isInterfacePointerNull)
```

Comment:
Actually, I suspect there's another bug right here: If a COM component does not implement a particular interface, `QueryInterface` is supposed to return a `HRESULT` of `E_NOINTERFACE` (0x80004002), not `S_OK` (0). Therefore I strongly suspect this check should really be:

```csharp
if (result == 0x80004002 && isInterfacePointerNull)
```

but I'd like to prove this with a test before I submit the fix.